gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsRecord.java

```diff
@@ -0,0 +1,270 @@
+  }
+
+  DnsRecord(Builder builder) {
+    this.name = builder.name;
+    this.rrdatas = ImmutableList.copyOf(builder.rrdatas);
+    this.ttl = builder.ttl;
+    this.type = builder.type;
+    this.zoneName = builder.zoneName;
+    this.zoneId = builder.zoneId;
+  }
+
+  /**
+   * Enum for the DNS record types supported by Cloud DNS.
+   *
+   * <p> Google Cloud DNS currently supports records of type A, AAAA, CNAME, MX NAPTR, NS, PTR, SOA,
+   * SPF, SRV, TXT.
+   *
+   * @see <a href="https://cloud.google.com/dns/what-is-cloud-dns#supported_record_types">Cloud DNS
+   * supported record types</a>
+   */
+  public enum DnsRecordType {
+    A,
+    AAAA,
+    CNAME,
+    MX,
+    NAPTR,
+    NS,
+    PTR,
+    SOA,
+    SPF,
+    SRV,
+    TXT;
+  }
+
+  public static class Builder {
+
+    private List<String> rrdatas = new LinkedList<>();
+    private String name;
+    private Integer ttl;
+    private DnsRecordType type;
+    private String zoneName;
+    private Long zoneId;
+
+    private Builder() {
+    }
+
+    /**
+     * Creates a builder and pre-populates attributes with the values from the provided DnsRecord
+     * instance.
+     */
+    public Builder(DnsRecord record) {
+      this.name = record.name;
+      this.ttl = record.ttl;
+      this.type = record.type;
+      this.zoneId = record.zoneId;
+      this.zoneName = record.zoneName;
+      this.rrdatas.addAll(record.rrdatas);
+    }
+
+    /**
+     * Adds a record to the record set. The records should be as defined in RFC 1035 (section 5) and
+     * RFC 1034 (section 3.6.1). Examples of records are available in Google DNS documentation.
+     *
+     * @see <a href="https://cloud.google.com/dns/what-is-cloud-dns#supported_record_types">Google
+     * DNS documentation </a>.
+     */
+    public Builder add(String record) {
+      this.rrdatas.add(checkNotNull(record));
+      return this;
+    }
+
+    /**
+     * Sets name for this DNS record set. For example, www.example.com.
+     */
+    public Builder name(String name) {
+      this.name = checkNotNull(name);
+      return this;
+    }
+
+    /**
+     * Sets the number of seconds that this record can be cached by resolvers. This number must be
+     * non-negative.
+     *
+     * @param ttl A non-negative number of seconds
+     */
+    public Builder ttl(int ttl) {
+      checkArgument(ttl >= 0, "TTL cannot be negative. The supplied value was " + ttl + ".");
+      this.ttl = ttl;
+      return this;
+    }
+
+    /**
+     * The identifier of a supported record type, for example, A, AAAA, MX, TXT, and so on.
+     */
+    public Builder type(DnsRecordType type) {
+      this.type = checkNotNull(type);
+      return this;
+    }
+
+    /**
+     * Builds the DNS record.
+     */
+    public DnsRecord build() {
+      return new DnsRecord(this);
+    }
+
+    /**
+     * Sets references to the managed zone that this DNS record belongs to.
+     */
+    public Builder managedZone(ManagedZoneInfo parent) {
```

Comment:
Is this method supposed to be used by the user? Or is it just for internal use?
To my understanding to create a DNS record we need to create a Change entity, change entity which must be associated to a particular ManagedZone. The user will have to specify the managed zone for Change, but there is no need for a user to also specify the managed zone in `DnsRecord`. If we want to keep this property we should make it package private. I suggest you try to implement the `toPb` and `fromPb` methods to get an understanding of which information is available and which not.

Also I think that we might need to change this to `public Builder managedZone(String zoneName)` as when [listing DNS records](https://cloud.google.com/dns/api/v1/resourceRecordSets/list) we do not have the id of the zone (nor a complete `ManagedZoneInfo`).